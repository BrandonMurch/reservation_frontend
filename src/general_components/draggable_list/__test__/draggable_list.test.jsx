// Dependencies
import React from 'react';
import PropTypes from 'prop-types';
import {
  render, fireEvent, screen, createEvent, waitFor,
} from '@testing-library/react';
import { create } from 'react-test-renderer';
import { act } from 'react-dom/test-utils';

// Components
import DraggableList from '../index';
import DisplayComponent from './test_display_component';

describe('<DraggableList />', () => {
  const props = {
    DisplayComponent,
    items: [{
      name: 'George',
      gender: 'male',
    }, {
      name: 'Sarah',
      gender: 'female',
    }],
    Headers: () => (
      <>
        <h2>Name</h2>
        <h2>Gender</h2>
      </>
    ),
    updateList: jest.fn(),
    getName: (person) => person.name,
    AddComponent: () => <button type="button">Add</button>,
  };
  beforeEach(() => {
    render(
      <DraggableList {...props} />,
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const tree = create(
      <DraggableList {...props} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should order items when dragged and dropped', async () => {
    const first = screen.getByText(/george/i);
    const second = screen.getByTestId(/endOfList/i);
    const dragStart = createEvent.dragStart(first);
    const dropEvent = createEvent.drop(second);
    Object.assign(dragStart, { dataTransfer: { setData: jest.fn() } });
    fireEvent(first, dragStart);
    Object.assign(dropEvent, { dataTransfer: { getData: jest.fn(() => 0) } });
    fireEvent(second, dropEvent);
    await waitFor(() => (
      expect(props.updateList)
        .toHaveBeenCalledWith([props.items[1], props.items[0]])
    ));
  });

  it('should order items when dragged and dropped in reverse', async () => {
    const first = screen.getByText(/sarah/i);
    const second = screen.getByText(/george/i);
    const dragStart = createEvent.dragStart(first);
    const dropEvent = createEvent.drop(second);
    Object.assign(dragStart, { dataTransfer: { setData: jest.fn() } });
    fireEvent(first, dragStart);
    Object.assign(dropEvent, { dataTransfer: { getData: jest.fn(() => 1) } });
    fireEvent(second, dropEvent);
    await waitFor(() => (
      expect(props.updateList)
        .toHaveBeenCalledWith([props.items[1], props.items[0]])
    ));
  });
});
