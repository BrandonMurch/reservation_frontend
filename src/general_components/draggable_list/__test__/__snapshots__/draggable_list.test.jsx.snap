// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DraggableList /> should match snapshot 1`] = `
<div
  className="TableContainer"
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <div
    className="Table"
  >
    <div>
      <h2>
        Name
      </h2>
      <h2>
        Gender
      </h2>
    </div>
    <ul>
      <li
        onDragOver={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
      />
      <li
        className="Row"
        data-position={0}
        draggable={true}
        id="table 0"
        onDragEnter={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "cursor": "pointer",
            "position": "relative",
          }
        }
      >
        <div>
          <h1>
            George
          </h1>
          <h1>
            male
          </h1>
        </div>
      </li>
      <li
        onDragOver={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
      />
      <li
        className="Row"
        data-position={1}
        draggable={true}
        id="table 1"
        onDragEnter={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "cursor": "pointer",
            "position": "relative",
          }
        }
      >
        <div>
          <h1>
            Sarah
          </h1>
          <h1>
            female
          </h1>
        </div>
      </li>
      <li
        onDragOver={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
      />
      <li
        className="Row"
        data-position={2}
        draggable={true}
        id="table 2"
        onDragEnter={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "cursor": "pointer",
            "position": "relative",
          }
        }
      >
        <div
          data-testid="endOfList"
          style={
            Object {
              "height": "2rem",
            }
          }
        />
      </li>
      <div
        className="Row"
      >
        <button
          type="button"
        >
          Add
        </button>
      </div>
    </ul>
  </div>
</div>
`;
